File: app/src/main/java/com/example/sergejmitrofanov/librarycatalogue/di/PresenterModule.java
package com.example.sergejmitrofanov.librarycatalogue.di;

import androidx.annotation.NonNull;
import com.example.sergejmitrofanov.librarycatalogue.presenter.BookListPresenter;
import com.example.sergejmitrofanov.librarycatalogue.presenter.BookListPresenterImpl;

class PresenterModule {

  @NonNull
  private final InteractorModule interactorModule;

  PresenterModule() {
    this.interactorModule = new InteractorModule();
  }

  BookListPresenter getFavoriteBooksPresenter() {
    return new BookListPresenterImpl(
        interactorModule.getFavoriteBooksUseCase()
    );
  }

  BookListPresenter getInternetBooksPresenter() {
    return new BookListPresenterImpl(
        interactorModule.getInternetBooksUseCase()
    );
  }

  BookListPresenter getForbiddenBooksPresenter() {
    return new BookListPresenterImpl(
        interactorModule.getForbiddenBooksUseCase()
    );
  }
}

File: app/src/main/java/com/example/sergejmitrofanov/librarycatalogue/di/PresenterModule.kt
package com.example.sergejmitrofanov.librarycatalogue.di

import com.example.sergejmitrofanov.librarycatalogue.interactor.BookListUseCase
import com.example.sergejmitrofanov.librarycatalogue.presenter.BookListPresenter
import com.example.sergejmitrofanov.librarycatalogue.presenter.BookListPresenterImpl
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ActivityComponent
import dagger.hilt.android.scopes.ActivityScoped

@Module
@InstallIn(ActivityComponent::class)
object PresenterModule {
    @Provides
    @ActivityScoped
    @FavoriteBooks
    fun provideFavoritePresenter(
        @FavoriteBooks useCase: BookListUseCase
    ): BookListPresenter = BookListPresenterImpl(useCase)


    @Provides
    @ActivityScoped
    @InternetBooks
    fun provideInternetPresenter(
        @InternetBooks useCase: BookListUseCase
    ): BookListPresenter = BookListPresenterImpl(useCase)

    @Provides
    @ActivityScoped
    @ForbiddenBooks
    fun provideForbiddenPresenter(
        @ForbiddenBooks useCase: BookListUseCase
    ): BookListPresenter = BookListPresenterImpl(useCase)
}
